modules/ttdBidAdapter.js

```diff
@@ -0,0 +1,602 @@
+  if (bannerParams || !(mediaTypesVideo || bannerParams)) {
+    mediaTypes[BANNER] = banner(bid);
+  } else if (mediaTypesVideo) {
+    mediaTypes[VIDEO] = video(bid);
+  }
+
+  Object.assign(impression, mediaTypes);
+
+  let bidfloor = getBidFloor(bid);
+  if (bidfloor) {
+    impression.bidfloor = parseFloat(bidfloor);
+    impression.bidfloorcur = 'USD';
+  }
+
+  return impression;
+}
+
+function getSizes(sizes) {
+  const sizeStructs = utils.parseSizesInput(sizes)
+    .filter(x => x) // sizes that don't conform are returned as null, which we want to ignore
+    .map(x => x.split('x'))
+    .map(size => {
+      return {
+        width: parseInt(size[0]),
+        height: parseInt(size[1]),
+      }
+    });
+
+  return sizeStructs;
+}
+
+function banner(bid) {
+  if (!(bid.mediaType === BANNER || utils.deepAccess(bid, 'mediaTypes.banner'))) {
+    return null;
+  }
+  const sizes = getSizes(bid.mediaTypes.banner.sizes).map(x => {
+    return {
+      w: x.width,
+      h: x.height,
+    }
+  });
+  const pos = parseInt(utils.deepAccess(bid, 'mediaTypes.banner.pos'));
+  const expdir = utils.deepAccess(bid, 'params.banner.expdir');
+  let optionalParams = {};
+  if (pos) {
+    optionalParams.pos = pos;
+  }
+  if (expdir) {
+    optionalParams.expdir = expdir;
+  }
+
+  const banner = Object.assign(
+    {
+      w: sizes[0].w,
+      h: sizes[0].h,
+      format: sizes,
+    },
+    optionalParams);
+  return banner;
+}
+
+function video(bid) {
+  const minduration = utils.deepAccess(bid, 'mediaTypes.video.minduration');
+  const maxduration = utils.deepAccess(bid, 'mediaTypes.video.maxduration');
+  const playerSize = utils.deepAccess(bid, 'mediaTypes.video.playerSize');
+  const api = utils.deepAccess(bid, 'mediaTypes.video.api');
+  const mimes = utils.deepAccess(bid, 'mediaTypes.video.mimes');
+  const placement = utils.deepAccess(bid, 'mediaTypes.video.placement');
+  const protocols = utils.deepAccess(bid, 'mediaTypes.video.protocols');
+  const playbackmethod = utils.deepAccess(bid, 'mediaTypes.video.playbackmethod');
+  const pos = utils.deepAccess(bid, 'mediaTypes.video.pos');
+  const startdelay = utils.deepAccess(bid, 'mediaTypes.video.startdelay');
+  const skip = utils.deepAccess(bid, 'mediaTypes.video.skip');
+  const skipmin = utils.deepAccess(bid, 'mediaTypes.video.skipmin');
+  const skipafter = utils.deepAccess(bid, 'mediaTypes.video.skipafter');
+  const minbitrate = utils.deepAccess(bid, 'mediaTypes.video.minbitrate');
+  const maxbitrate = utils.deepAccess(bid, 'mediaTypes.video.maxbitrate');
+
+  let video = {
+    minduration: minduration,
+    maxduration: maxduration,
+    api: api,
+    mimes: mimes,
+    placement: placement
+  };
+
+  if (utils.isArray(playerSize[0])) {
+    video.w = parseInt(playerSize[0][0]);
+    video.h = parseInt(playerSize[0][1]);
+  } else if (utils.isNumber(playerSize[0])) {
+    video.w = parseInt(playerSize[0]);
+    video.h = parseInt(playerSize[1]);
+  }
+
+  if (protocols) {
```

Comment:
You reject bids missing protocols below so you should be able to just set it above without the need of checking again

Reply:
fixed